### DISCUSSION TITLE
Claw Import / Export Syntax

### DISCUSSION THREAD

**Discussion by esoterra** [esoterra/claw-lang#10] [General]
## Starting Point

My initial design for function definition and export looked like this
```js
func foo(a: u32) -> u32 { ... }

export func bar(a: u32) -> u32 { ... }
```
but when combined with the import syntax it looks somewhat out of place.
```js
// Option A.
import imported: func(a: u32) -> u32;

export func exported(a: u32) -> u32 {
    return imported(a);
}
```

## Alternatives

It may be better to make it mirror the import syntax more directly. The symmetry certainly looks nice.
```js
// Option B
import imported: func(a: u32) -> u32;

export exported: func(a: u32) -> u32 {
    return imported(a);
}
```

On the other hand this means that `a: b`, which previously always meant `a` has type `b`, can now mean `a` is defined as `b`.

One way to make this more distinct is to introduce an `=` and potentially ending `;` to make it clearly a definition/assignment.

```js
// Option C
import imported: func(a: u32) -> u32;

export exported: func(a: u32) -> u32 = {
    return imported(a);
}

// Option D - C but with a semicolon at the end of the export item
```

```js
// Option E
import imported: func(a: u32) -> u32;

export exported = func(a: u32) -> u32 {
    return imported(a);
}

// Option F - E but with a semicolon at the end of the export item
```

## Value Import / Export

Another thing to consider is that if/when value imports arrive, we'll probably want function and value import and exports to all make sense together.

It's my understanding that value imports/exports in WIT will probably look like this:
```js
import val-in: u32;
export val-out: u32;
```
<details>

<summary>Examples with Values</summary>
Here's my best guess of how we'd introduce value imports and exports into the examples above.

```js
// Option A.
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func func-out(a: u32) -> u32 {
    return func-in(a);
}
```

```js
// Option B
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func-out: func(a: u32) -> u32 {
    return func-in(a);
}
```

```js
// Option C
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func-out: func(a: u32) -> u32 = {
    return func-in(a);
}
```

```js
// Option D
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func-out: func(a: u32) -> u32 = {
    return func-in(a);
};
```

```js
// Option E
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func-out = func(a: u32) -> u32 {
    return func-in(a);
}
```

```js
// Option F
import val-in: u32;
import func-in: func(a: u32) -> u32;

export val-out: u32 = 2;
export func-out = func(a: u32) -> u32 {
    return func-in(a);
};
```
</details>

## Call for Feedback

If you have any additional ideas, feel free to provide them below.
If you think we should go with a specific option, please comment below and include your reasoning.

**Comment by esoterra** [comment]
# Current Thoughts

I'm currently leaning towards C & D because they have the strongest symmetry and also enable some pretty cool options for functions whose bodies are expressions.

e.g.
```js
export func-out: func(a: u32) -> u32 = a * 2;
```

```js
import func-in: func(a: u32) -> u32;

export func-out: func(a: u32) -> u32 = func-in(a);
```

I guess there's nothing preventing this syntax from being legal with option B, for example, it just feels more consistent with C & D.

**Comment by lann** [comment]
I'd lean toward Option B for the purely selfish reason of not getting frustrated when I switch from writing Rust, forget a `=` in a func export, then swear at the screen.

**Reply by esoterra** [reply]
Also a few people have argued that the simple case (defining an exported function) should ideally use less punctuation where possible, and C/D do have an extra `=`/`;` which people will need to remember and type each time.

**Reply by esoterra** [reply]
I've implemented Option B for now in #11

**Comment by lukewagner** [comment]
FWIW, I like Option A the most.  From a C-M perspective, when you write `export func foo(...) { ... }`, it's syntactic sugar for doing 2 things:
* `func foo(...) { ... }`
* `export foo;`

Thus, Option A is attractive b/c it's a trivial folding of these two definitions doesn't need any more invasive surgery to the syntax of function definitions.  The fact that imports and exports look asymmetric follows from the fact that, in both CLAW and the underlying C-M, export definitions are asymmetric with import definitions in that they aren't forced to re-state the type (the principle type of the exported definition is used).  It's only inside component *types* (and thus `world`s) that import and export *declarators* (our distinct term for the "the thing you have a list of inside a component/module/instance type") get to be symmetric.

Oh, also, and I know this won't be super-compelling to some... Option A is symmetric with JS module export syntax :P

**Reply by esoterra** [reply]
I've been interpreting B as being non-invasive with the function definition syntax in much the way you're saying, but at the cost of making a standalone function definition look a bit strange.

```js
foo: func(...) {
   ...
}
```
or
```js
let foo: func(...) {
   ...
}
```

Thinking about it a bit more, neither of these look that great on their own and to your point making them look more sensible would make them less consistent with exported functions.

More generally and to your other point, A is the most consistent with the majority of programming languages.
